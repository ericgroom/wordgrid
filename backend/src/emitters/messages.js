exports.chatMessage = (io, gameId, message, sender) => {
  io.of("/game")
    .in(`${gameId}`)
    .emit("chat message", { message, sender: sender.nickname });
};
