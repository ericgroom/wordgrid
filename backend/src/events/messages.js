const emitters = require("../emitters");
const utils = require("./utils");
const { decodeId } = require("../utils");

exports.onChatMessage = async (io, socket, message, gameId) => {
  const user = await utils.getCurrentUser(socket, gameId);
  if (user.nickname) {
    emitters.messages.chatMessage(io, gameId, message, user);
  }
};

exports.registerListeners = (io, socket) => {
  socket.on("send chat message", async ({ message, gameId }) => {
    gameId = decodeId(gameId);
    console.log(`${socket.id} sends message: ${message}`);
    await exports.onChatMessage(io, socket, message, gameId);
  });
};
