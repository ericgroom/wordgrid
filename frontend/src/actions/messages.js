export const RECIEVED_MESSAGE = "RECIEVED_MESSAGE";
export const SEND_MESSAGE = "SEND_MESSAGE";

export const recievedMessage = message => ({ type: RECIEVED_MESSAGE, message });
