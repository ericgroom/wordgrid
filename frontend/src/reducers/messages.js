import { LEAVE_GAME, RECEIVED_MESSAGE } from "../actions";

const initialState = {
  messages: [],
  messageId: 0
};

export default (state = initialState, action) => {
  switch (action.type) {
    case RECEIVED_MESSAGE:
      return {
        ...state,
        messages: [
          ...state.messages,
          { ...action.message, id: state.messageId }
        ],
        messageId: state.messageId + 1
      };
    case LEAVE_GAME:
      return initialState;
    default:
      return state;
  }
};
