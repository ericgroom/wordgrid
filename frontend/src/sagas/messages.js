import io from "socket.io-client";
import { eventChannel } from "redux-saga";
import { take, call, all } from "redux-saga/effects";
import { SEND_MESSAGE, recievedMessage } from "../actions";
import { putFrom } from "./index";

/**
 * Listens for specific actions and handles them. These should be limited to
 * actions that need to emit to a socket.
 * @param {*} socket socket.io instance
 */
function* messageActionListener(socket) {
  while (true) {
    const action = yield take([SEND_MESSAGE]);
    switch (action.type) {
      case SEND_MESSAGE:
        socket.emit("chat message", action.message);
        break;
      default:
        throw new Error(
          `messageActionListener saga took action but has no handler for: ${
            action.type
          }`
        );
    }
  }
}

/**
 * Listens to chat socket and emits actions based on socket events
 * @param {*} socket socket.io socket instance
 */
function messageListener(socket) {
  return eventChannel(emit => {
    socket.on("chat message", msg => {
      return emit(recievedMessage(msg));
    });
    return () => {
      console.log("closing socket");
      socket.close();
    };
  });
}

export default function* messagesFlow() {
  while (true) {
    const socket = io.connect("http://localhost:3001/chat");
    const socketChannel = yield call(messageListener, socket);
    yield all([putFrom(socketChannel), messageActionListener(socket)]);
  }
}
