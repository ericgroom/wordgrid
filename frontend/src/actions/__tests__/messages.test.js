import { receivedMessage, RECEIVED_MESSAGE } from "../messages";

describe("messages actions", () => {
  it("receivedMessage", () => {
    expect(receivedMessage("hello")).toMatchSnapshot();
  });
});
