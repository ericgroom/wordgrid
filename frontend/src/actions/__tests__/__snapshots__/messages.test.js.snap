// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`messages actions receivedMessage 1`] = `
Object {
  "message": "hello",
  "type": "RECEIVED_MESSAGE",
}
`;
